src_c/surface.c

```diff
@@ -2105,6 +2108,300 @@ surf_blits(pgSurfaceObject *self, PyObject *args, PyObject *keywds)
     return RAISE(PyExc_TypeError, "Unknown error");
 }
 
+#define UBLITS_ERR_TUPLE_REQUIRED 11
+#define UBLITS_ERR_INSUFFICIENT_ARGS 12
+#define UBLITS_ERR_FLAG_NOT_NUMERIC 13
+#define UBLITS_ERR_DORETURN_NOT_NUMERIC 14
+#define UBLITS_SERR_LISTORTUPLE_REQUIRED \
+    "blit_sequence can only be a list or a tuple"
+static PyObject *
+surf_ublits(pgSurfaceObject *self, PyObject *const *args, Py_ssize_t nargs)
+{
+    SDL_Surface *src, *dest = pgSurface_AsSurface(self);
+    SDL_Rect *src_rect, temp, dest_rect;
+
+    PyObject *srcobject = NULL, *argpos = NULL;
+    PyObject *blitsequence, *tmpblitseq;
+    PyObject *iterator = NULL;
+    PyObject *item = NULL;
+    PyObject *ret = NULL;
+    PyObject *retrect = NULL;
+    PyObject **f_blitsequence;
+    Py_ssize_t itemlength, sequencelength, seq_counter;
+    int doreturn = 1;
+    int errornum = 0;
+    int result;
+    int sx, sy;
+    int flags_numeric;
+
+    temp.x = 0;
+    temp.y = 0;
+
+    if (nargs < 3) {
+        errornum = UBLITS_ERR_INSUFFICIENT_ARGS;
+        goto on_error;
+    }
+
+    blitsequence = args[0];
+    if (!PyLong_Check(args[1]) || !pg_IntFromObj(args[1], &flags_numeric)) {
+        errornum = UBLITS_ERR_FLAG_NOT_NUMERIC;
+        goto on_error;
+    }
+
+    if (!pg_IntFromObj(args[2], &doreturn)) {
+        errornum = UBLITS_ERR_DORETURN_NOT_NUMERIC;
+        goto on_error;
+    }
+
+    if (!dest) {
+        errornum = BLITS_ERR_DISPLAY_SURF_QUIT;
+        goto on_error;
+    }
+
+    /* Generator */
+    if (!PyList_Check(blitsequence) && !PyTuple_Check(blitsequence)) {
```

Comment:
Not sure whether this block of code is worth the extra complexity here. 
`PySequence_Fast` should be able to handle most, if not all iterators

Reply:
Yes I'm aware that it's not ideal. I tried using the return value from PySequence_Fast but I couldn't get it to correctly distinguish between a generator and a tuple/list. I'll try again because I agree this must change.

Reply:
And i don't get how PySequence_Fast  treats errors because the docs say it 

> raises [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) with m as the message text. Returns NULL on failure.
So what does the function do? shouldn't the program end with a TypeError? i don't get what happens

Reply:
The function basically returns `NULL`, and on our side we don't need to set a new error again. We only need to pass the message string to the function as argument `m`

Reply:
The big problem is that PySequence_Fast works with all iterables so you can't get whether it's a generator or a list/tuple being passed.